From 10e079a7fd8b2b162c2fa3ac34522361a0eafc9c Mon Sep 17 00:00:00 2001
From: Sultanxda <sultanxda@gmail.com>
Date: Sat, 3 Feb 2018 22:31:16 -0800
Subject: [PATCH] Force HAL3 mode for front-facing camera

HAL1 mode for the front-facing camera is buggy. Videos recorded in HAL1
mode with the front-facing camera (using any app) come out garbled for
some reason. However, HAL3 mode works fine, so force HAL3 mode everywhere
for the front-facing camera.

This also retains OnePlus' fix for HAL3/HAL1 coexistence with the rear
camera.

Change-Id: Ib4964e21ce39a5fa6f7f142f78cf4e0b75b65c78
---
 services/camera/libcameraservice/CameraService.h | 9 +++++++++
 1 file changed, 9 insertions(+)

diff --git a/services/camera/libcameraservice/CameraService.h b/services/camera/libcameraservice/CameraService.h
index 25c109aa8..0337821fa 100644
--- a/services/camera/libcameraservice/CameraService.h
+++ b/services/camera/libcameraservice/CameraService.h
@@ -831,6 +831,15 @@ binder::Status CameraService::connectHelper(const sp<CALLBACK>& cameraCb, const
 
     int originalClientPid = 0;
 
+    if (cameraIdToInt(cameraId) == 1) {
+        // Force HAL3 mode for front camera since HAL1 mode is buggy
+        halVersion = CAMERA_DEVICE_API_VERSION_3_4;
+    } else if (effectiveApiLevel == API_1 &&
+                halVersion == CAMERA_HAL_API_VERSION_UNSPECIFIED) {
+        // If a HAL version isn't assigned with API1, then set HAL1 by default
+        halVersion = CAMERA_DEVICE_API_VERSION_1_0;
+    }
+
     ALOGI("CameraService::connect call (PID %d \"%s\", camera ID %s) for HAL version %s and "
             "Camera API version %d", clientPid, clientName8.string(), cameraId.string(),
             (halVersion == -1) ? "default" : std::to_string(halVersion).c_str(),
-- 
2.14.1

